freemarker-core/src/main/javacc/freemarker/core/FTL.jj

```diff
@@ -3892,18 +3895,50 @@ SwitchBlock Switch() :
     }
     [
 	    (
-	        caseIns = Case()
-	        {
-	            if (caseIns.condition == null) {
-	                if (defaultFound) {
-	                    throw new ParseException(
-	                    "You can only have one default case in a switch statement", template, start);
-	                }
-	                defaultFound = true;
-	            }
-	            switchBlock.addCase(caseIns);
-	        }
-	    )+
+            (
+                caseIns = Case()
+                {
+                    if (caseIns.condition == null) {
+                        if (defaultFound) {
+                            throw new ParseException(
+                            "You can only have one default case in a switch statement", template, start);
+                        }
+                        defaultFound = true;
+                    }
+                    switchBlock.addCase(caseIns);
+                }
+            )+
+            |
+            (
+                {
+                    // A Switch with Case supports break, but not one with On.
+                    // Do it this way to ensure backwards compatibility.
+                    breakableDirectiveNesting--;
```

Comment:
Will that also work when switch-case is nested inside switch-on?

Reply:
A few lines up, at the SwitchBlock level, there's an increment. As soon as we notice we're in the `on` case, then we undo the increment. If there was a nested switch-case then it would be incremented again. This counter is only used at parse time to determine if we're in a context where `break` is allowed. Because `on` doesn't allow a closing tag, I don't think there's a possibility for a disallowed break to accidentally be allowed between `on` tags.